plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/graphvisitors/InteractiveLayeredGraphVisitor.java

```diff
@@ -0,0 +1,451 @@
+
+/**
+ * Graph visitor which visits only the root node and recursively steps through the graph
+ * to set interactive option for the layered algorithm.
+ */
+public class InteractiveLayeredGraphVisitor implements IGraphElementVisitor {
+
+    /**
+     * Constant help with node placement, since nodes should not overlap and the their width and height may not be
+     * final.
+     */
+    public static final int NODE_PLACEMENT_HELPER = 100000;
+    
+    /**
+     * Set as the layer of a node that should be in the last layer.
+     */
+    public static final int LAST_LAYER = 100000;
+
+    /*
+     * (non-Javadoc)
+     * 
+     * @see org.eclipse.elk.core.util.IGraphElementVisitor#visit(org.eclipse.elk.graph.ElkGraphElement)
+     */
+    @Override
+    public void visit(final ElkGraphElement element) {
+        // Only apply to root of the graph
+        if (element.eContainer() == null && element instanceof ElkNode) {
+            ElkNode root = (ElkNode) element;
+            setInteractiveOptions(root);
+        }
+    }
+
+    /**
+     * Sets the coordinates of the nodes in the graph.
+     * 
+     * @param root
+     *            Root of the graph
+     */
+    private void setInteractiveOptions(final ElkNode root) {
+        String algorithm = root.getProperty(CoreOptions.ALGORITHM);
+        if (algorithm == null || LayeredOptions.ALGORITHM_ID.endsWith(algorithm)) {
+            // Make sure that potential usages of the LayerConstraint are translated into
+            // LayerChoiceConstraints
+            for (ElkNode node : root.getChildren()) {
+                if (node.hasProperty(LayeredOptions.LAYERING_LAYER_CONSTRAINT)) {
+                    LayerConstraint constraint = node.getProperty(LayeredOptions.LAYERING_LAYER_CONSTRAINT);
+                    node.setProperty(LayeredOptions.LAYERING_LAYER_CONSTRAINT, LayerConstraint.NONE);
+                    switch (constraint) {
+                    case FIRST:
+                        if (node.getProperty(LayeredOptions.LAYERING_LAYER_CHOICE_CONSTRAINT) == -1) {
+                            node.setProperty(LayeredOptions.LAYERING_LAYER_CHOICE_CONSTRAINT, 0);
+                        }
+                        break;
+                    case LAST:
+                        if (node.getProperty(LayeredOptions.LAYERING_LAYER_CHOICE_CONSTRAINT) == -1) {
+                            node.setProperty(LayeredOptions.LAYERING_LAYER_CHOICE_CONSTRAINT, LAST_LAYER);
+                        }
+                        break;
+                    }
+                }
+            }
+        }
+        for (ElkNode n : root.getChildren()) {
+            if (!n.getChildren().isEmpty()) {
+                setInteractiveOptions(n);
+            }
+        }
+        if (algorithm == null || LayeredOptions.ALGORITHM_ID.endsWith(algorithm)) {
+            if (!root.getChildren().isEmpty()) {
+                setCoordinates(root);
+                setInteractiveStrategies(root);
+            }
```

Comment:
Why is this not part of the first `if`?

Reply:
I guess due to legacy reasons before I was sure that it is not possible to correctly estimate the width and height of a node. I will include it in the first `if`